Macro Expansion Producing Expression with Semicolon

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
macro_rules! mac {
    () => {return 1;}
}

pub fn foo() -> i32 {
    mac!()
}
```

I expected to see this happen:
```
warning: trailing semicolon in macro used in expression position
 --> src/lib.rs:2:20
  |
2 |     () => {return 1;}
  |                    ^
...
6 |     mac!()
  |     ------ in this macro invocation
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, [see issue #79813 <https://github.com/rust-lang/rust/issues/79813>](https://github.com/rust-lang/rust/issues/79813)
  = note: macro invocations at the end of a block are treated as expressions
  = note: to ignore the value produced by the macro, add a semicolon after the invocation of `mac`
  = note: `#[warn(semicolon_in_expressions_from_macros)]` on by default
  = note: this warning originates in the macro `mac` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Instead, this happened:
```
test.rs:2:20: error: tokens here and after are unparsed
    2 |     () => {return 1;}
      |                    ^
```

### Meta

- git sha: ```1b322540fa0fce561cf86e371dd7f304f1343511```

--- a/gcc/rust/expand/rust-macro-expand.cc
+++ b/gcc/rust/expand/rust-macro-expand.cc
@@ -911,6 +911,15 @@ transcribe_expression (Parser<MacroInvocLexer> &parser)
   if (expr == nullptr)
     return AST::Fragment::create_error ();
 
+  // FIXME: make this an error for some edititons
+  if (parser.peek_current_token ()->get_id () == SEMICOLON)
+    {
+      rust_warning_at (
+	parser.peek_current_token ()->get_locus (), 0,
+	"trailing semicolon in macro used in expression context");
+      parser.skip_token ();
+    }
+
   auto end = lexer.get_offs ();
 
   return AST::Fragment ({std::move (expr)}, lexer.get_token_slice (start, end));

--- a/gcc/testsuite/rust/compile/macro-issue2273.rs
+++ b/gcc/testsuite/rust/compile/macro-issue2273.rs
@@ -1,0 +1,7 @@
+macro_rules! mac {
+    () => {();} // { dg-warning "trailing semicolon" }
+}
+
+pub fn foo() {
+    mac!()
+}

